### ISSUE TITLE
3.3.1-enterprise - Returning results from relationships that don't exist

### ISSUE THREAD

**Issue by senica** [neo4j/neo4j#10611]
I posted this in slack. I ended up downgrading to 3.2.8 and everything works as expected. I don't have a good/short way to reproduce. I would be happy to set everything back up to 3.3.1 and show someone the error if it helped, but I mostly wanted to make this issue in case anyone else experiencing something similar there is a larger consensus.

My query - ran either direclty in the neo4j browser, or with the js driver.
```
rows = await _db.run(`MATCH (u:user {uuid: $uid})-[:allowed_to]->(c:company {uuid: $cid})
MATCH (f:file {uuid: $uuid})
MATCH (f)-[:uploaded_for]->(t) // get targets
OPTIONAL MATCH (tg:tag)-[:tags]->(f) // get tags
RETURN {
  uuid: f.uuid,
  targets: collect(DISTINCT t),
  tags: collect(DISTINCT tg)
}
`, {
uid: req.user.uid,
cid: req.user.cid,
uuid: fid,
})
```
What seems to be happening is that it's throwing things into `tags` that don't have the relationship of `tags`.
For example, a partial output of the results is (everything else is as expected):
```
...
"tags": [{
  "identity": {
  "low": 3132,
  "high": 0
},
"labels": [
  "note"
],
"properties": {
  "note": "What is this?",
  "updated_at": 1512661314571,
  "editable": false,
  "created_at": 1512661314571,
  "details": "Is this working right?",
  "label": "dooda",
  "category": "mycat",
  "uuid": "2bca4acb-0241-4a4a-ae37-ba34f92b3acd"
}}, {
"identity": {
  "low": 20,
  "high": 0
},
"labels": [
 "company"
],
"properties": {
  "name": "Allebrum",
  "uuid": "45a8404a-073a-49c3-8705-63142e31ff5d",
  "email": ""
}}]}
```
but if I look at the `tags` relationship, all I have is one relationship that has nothing to do with the ones specified: 
![screen shot 2017-12-07 at 12 15 09 pm](https://user-images.githubusercontent.com/954952/33776649-ec6cde3e-dbf6-11e7-84e8-dc88f85dc568.jpg)

As mentioned, during development we used 3.2.5. We are using docker, so the installs on the production server we clean installs. At one point I had upgraded from 3.3.0 to 3.3.1 to fix an issue we were having that was resolved in 3.3.1. Thinking this was a bad upgrade, we deleted everything and started again...the same result happened.

- Neo4j version: 3.3.1-enterprise
- Operating system: Neo4j Docker 3.3.1-enterprise
- API/Driver: Cypher and JS neo4j driver produced same results
- **Steps to reproduce**

- Expected behavior
A single result.
- Actual behavior
Two results, with unreleated nodes placed in tags array

**Comment by Lojjs** [comment]
Thanks for reporting, this definitely sounds like a bug. We in the Cypher team will look at it as soon as possible.

**Comment by pontusmelke** [comment]
Thanks for reporting @senica 

I'll look at it - however if you have the opportunity to provide the execution plan it would be great. Just run the query with EXPLAIN or PROFILE prepended and paste the plan here. Another thing you could try is to run the query but with `CYPHER runtime=interpreted` prepended to it and see if that makes any difference.

Regards,
Pontus

**Comment by senica** [comment]
@pontusmelke I'll try and replicate locally today and give you the Profile results.

**Comment by senica** [comment]
@pontusmelke Here's the info you requested

Here is the entire query...just in case :)
```
EXPLAIN
MATCH (u:user {uuid: 'be104c68-8b63-42fa-8310-808b747917da'})-[:allowed_to]->(c:company {uuid: 'ce5fcf98-8413-407d-8624-09921ec2dec6'})
      MATCH (f:file {uuid: 'f26fa8bd-4559-47e5-b6d5-848620d37975'})
      MATCH (f)-[:uploaded_for]->(t) // get targets
      OPTIONAL MATCH (tg:tag)-[:tags]->(f) // get tags
      RETURN {
        uuid: f.uuid,
        acl: f.acl,
        title: f.title,
        description: f.description,
        original_name: f.original_name,
        mimetype: f.mimetype,
        encoding: f.encoding,
        size: f.size,
        created_at: f.created_at,
        updated_at: f.updated_at,
        targets: collect(DISTINCT t),
        tags: collect(DISTINCT tg)
        // bucket: f.bucket, internal use
        // bucket_key: f.bucket_key internal use
        // path: f.path, path is for internal use
      }
```
With EXPLAIN:
![plan](https://user-images.githubusercontent.com/954952/34280729-cc968dd0-e66e-11e7-8c35-fa9842c844e2.png)

WITH PROFILE:
![plan 2](https://user-images.githubusercontent.com/954952/34280768-09d685d8-e66f-11e7-9bee-76fa97cbeea2.png)

Running with `CYPHER runtime=interpreted` returns the correct text result (no tags and the correct targets). The graphical graph in the browser I'm not sure gives the correct result.
Here is the text result:
![screen shot 2017-12-21 at 4 57 07 pm](https://user-images.githubusercontent.com/954952/34280935-073596ce-e670-11e7-89cd-6dcabd92f79d.jpg)
Here is what the graph shows by the way (though that's not important to me), just fyi. I would expect to see a file in there somewhere:
![screen shot 2017-12-21 at 4 58 09 pm](https://user-images.githubusercontent.com/954952/34280950-252ce01a-e670-11e7-995d-d5ede0d55463.jpg)
Here is the explain for the interpreted runtime:
![plan 4](https://user-images.githubusercontent.com/954952/34280986-589f3ec0-e670-11e7-90c2-fc18c9b9b143.png)


Let me know what you think; if I can help further or if this is an error on my part.

**Comment by pontusmelke** [comment]
Thanks!
No this looks like a bug in the new default runtime, that is why it works when you force the database to use the old interpreted runtime using `runtime=interpreted`. 

Will look into it as soon as possible.

**Comment by fickludd** [comment]
This will be solved by #10749. Thank you for reporting @senica, and please confirm that this solves your issue in `3.3.2`.
